src/test/java/de/tum/in/www1/artemis/FileIntegrationTest.java

```diff
@@ -0,0 +1,243 @@
+
+    @Autowired
+    QuizQuestionRepository quizQuestionRepository;
+
+    @Autowired
+    ProgrammingExerciseRepository programmingExerciseRepository;
+
+    @Autowired
+    ProgrammingExerciseService programmingExerciseService;
+
+    @Autowired
+    FileService fileService;
+
+    @Autowired
+    ResultRepository resultRepo;
+
+    @Autowired
+    ParticipationService participationService;
+
+    @Autowired
+    LectureRepository lectureRepo;
+
+    @Autowired
+    private BambooRequestMockProvider bambooRequestMockProvider;
+
+    @Autowired
+    private BitbucketRequestMockProvider bitbucketRequestMockProvider;
+
+    @BeforeEach
+    public void initTestCase() {
+        database.addUsers(2, 2, 1);
+    }
+
+    @AfterEach
+    public void tearDown() {
+        database.resetDatabase();
+    }
+
+    @Test
+    @WithMockUser(value = "instructor1", roles = "INSTRUCTOR")
+    public void testSaveTempFile() throws Exception {
+        MockMultipartFile file = new MockMultipartFile("file", "file.png", "application/json", "some data".getBytes());
+        JsonNode response = request.postWithMultipartFile("/api/fileUpload?keepFileName=false", file.getOriginalFilename(), "file", file, JsonNode.class, HttpStatus.CREATED);
+        String responsePath = response.get("path").asText();
+
+        String responseFile = request.get(responsePath, HttpStatus.OK, String.class);
+        assertThat(responseFile).isEqualTo("some data");
+    }
+
+    @Test
+    @WithMockUser(value = "instructor1", roles = "INSTRUCTOR")
+    public void testGetTemplateFile() throws Exception {
+        database.addCourseWithOneProgrammingExerciseAndTestCases();
+        ProgrammingExercise programmingExercise = programmingExerciseRepository.findAllWithEagerParticipations().get(0);
+        database.addStudentParticipationForProgrammingExercise(programmingExercise, "student1");
+        database.addStudentParticipationForProgrammingExercise(programmingExercise, "student2");
+
+        java.io.File localRepoFile = Files.createTempDirectory("repo").toFile();
+        Git localGit = Git.init().setDirectory(localRepoFile).call();
+
+        java.io.File originRepoFile = Files.createTempDirectory("repoOrigin").toFile();
+        Git remoteGit = Git.init().setDirectory(originRepoFile).call();
+        StoredConfig config = localGit.getRepository().getConfig();
+        config.setString("remote", "origin", "url", originRepoFile.getAbsolutePath());
+        config.save();
+        doReturn(new GitUtilService.FileRepositoryUrl(originRepoFile)).when(versionControlService).getCloneRepositoryUrl(anyString(), anyString());
+
+        programmingExercise.setId(null);
+        request.post(ROOT + SETUP, programmingExercise, HttpStatus.OK);
+        // programmingExerciseService.setupProgrammingExercise(programmingExercise);
+
+        request.get("/files/templates/" + programmingExercise.getProgrammingLanguage().toString().toLowerCase() + "/exercise", HttpStatus.OK, byte[].class);
+    }
```

Comment:
This method throws an error while setting up the bitbucket project (line 132/133): **org.springframework.web.client.HttpClientErrorException$Unauthorized: 401**

What might I do wrong here? Is there an easier way to upload/create a template file, to test the GET request?

Reply:
I entered my credentials in the application-artemis.yml using my personal account, are there any special rights needed to setup these repositories?

Reply:
Yes, you need special permissions for those actions. I guess we want to mock bitbucket @krusche?

Reply:
I found these functions for mocking:
```
bitbucketRequestMockProvider.enableMockingOfRequests();
bitbucketRequestMockProvider.mockCreateProjectForExercise(programmingExercise);
bitbucketRequestMockProvider.mockCreateRepository(programmingExercise, "test");
```
Are these creating template files? If yes where and under which filename?

Reply:
It depends what you want to achieve here. If you simply need a programming exercise in the database, just add it directly using the corresponding repository.

If you need git repositories, you can also create them in the temp folder of the system and tell the git service (using mockito) to use them instead of the real ones.

You might also want to use these methods to mock bitbucket, but it really depends what you want to achieve here

Reply:
The main thing I want to achieve is to create a file in the folder 
`files/templates`
Unfortunately it is not possible to move files there manually using the `fileService` and the only way (I found) to create a file there is via the `setupProgrammingExercise` method. 
But it is not clear for me how to mock that method or whats the cleanest way to create such a file.

Reply:
then I suggest to leave this out for now. It does not make sense to call `setupProgrammingExercise` just to create files in the `templates` folder!

Reply:
Ok I disabled the test case and added a TODO.